src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -414,7 +422,266 @@ public static unsafe void Copy(Array sourceArray, int sourceIndex, Array destina
             }
 
             // Less common
-            CopyImpl(sourceArray!, sourceIndex, destinationArray!, destinationIndex, length, reliable: false);
+            CopyImpl(sourceArray, sourceIndex, destinationArray, destinationIndex, length, reliable: false);
+        }
+
+        private static unsafe void CopyImpl(Array? sourceArray, int sourceIndex, Array? destinationArray, int destinationIndex, int length, bool reliable)
+        {
+            ArgumentNullException.ThrowIfNull(sourceArray);
+            ArgumentNullException.ThrowIfNull(destinationArray);
+
+            if (sourceArray.GetType() != destinationArray.GetType() && sourceArray.Rank != destinationArray.Rank)
+                throw new RankException(SR.Rank_MustMatch);
+
+            ArgumentOutOfRangeException.ThrowIfNegative(length);
+
+            int srcLB = sourceArray.GetLowerBound(0);
+            ArgumentOutOfRangeException.ThrowIfLessThan(sourceIndex, srcLB);
+            if (sourceIndex - srcLB < 0)
+                throw new ArgumentException(SR.Arg_LongerThanSrcArray, nameof(sourceArray));
+            sourceIndex -= srcLB;
```

Comment:
```suggestion
            int adjustedSourceIndex = sourceIndex - srcLB;
            if ((adjustedSourceIndex < 0) && (uint)(adjustedSourceIndex + length) > sourceArray.NativeLength))
                throw new ArgumentException(SR.Arg_LongerThanSrcArray, nameof(sourceArray));
            sourceIndex = adjustedSourceIndex;
```
We can deduplicate the code a bit.

(Same for destinationIndex.)